File: Softeer/장애물 인식 프로그램.py
from sys import stdin
from collections import deque

n = int(stdin.readline())
graph = [[] for _ in range(n)]
visited = [[0] * n for _ in range(n)]

for i in range(n):
    graph[i] = list(map(int, stdin.readline().rstrip()))

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

res = []
def bfs(a, b):
    q = deque()
    q.append([a, b])
    cnt = 1
    visited[a][b] = 1

    while q:
        x, y = q.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0<=nx<n and 0<=ny<n and not visited[nx][ny] and graph[nx][ny] == 1:
                q.append([nx, ny])
                cnt += 1
                visited[nx][ny] = 1

    return cnt
                
        

for x in range(n):
    for y in range(n):
        if graph[x][y] == 1 and not visited[x][y]:
            res.append(bfs(x, y))
print(len(res))
for val in sorted(res):
    print(val)

File: Softeer/장애물 인식 프로그램.js
var fs = require('fs')
var input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');
var n = Number(input[0])
input.splice(0, 1)

var graph = new Array();
for(let i=0;i<n;i++){
    var lst = input[i].split('').map(Number);
    graph.push(lst)
}
var visited = Array.from(Array(n), () => Array(n).fill(0))


const dx = [-1, 1, 0, 0]
const dy = [0, 0, -1, 1]

const bfs = (a, b) => {
    var q = [[a, b]];
    visited[a][b] = 1;
    const value = graph[a][b];
    var tmp = 1;
    while(q.length > 0){
        var [x, y] = q.shift();

        for(let i = 0; i < 4; i++){
            var nx = x + dx[i];
            var ny = y + dy[i];
            if(0<=nx && nx < n && 0<=ny && ny<n && visited[nx][ny] === 0 && graph[nx][ny] === value){
                q.push([nx,ny])
                visited[nx][ny] = 1;
                tmp += 1;
            }
        }
    }
    return tmp;
}

var cnt = 0; // 총 블록 수
var res = new Array();
for(let x = 0; x < n; x++){
    for(let y = 0; y < n; y++){
        if(visited[x][y] === 0 && graph[x][y] === 1){
            cnt++;
            res.push(bfs(x, y));
        }
    }
}

console.log(cnt);
res.sort((a, b) => a - b);
res.forEach((e) => {
    console.log(e)
})